actionpack/lib/action_dispatch/middleware/host_authorization.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require "action_dispatch/http/request"
+
+module ActionDispatch
+  # This middleware guards from DNS rebinding attacks by white-listing the
+  # hosts a request can be sent to.
+  #
+  # When a request comes to an unauthorized host, the +response_app+
+  # application will be executed and rendered. If no +response_app+ is given, a
+  # default one will run, which responds with +403 Forbidden+.
+  class HostAuthorization
+    class SpoofProtection # :nodoc:
+      Error = Class.new(StandardError)
+
+      def self.detect_host(request)
+        new(request).detect_host
+      end
+
+      def initialize(request)
+        @request = request
+      end
+
+      def detect_host
+        raise Error if spoofed?(forwarded_host)
+
+        @request.host
+      end
+
+      private
+
+        # Special-Use Domain Names as defined in https://tools.ietf.org/html/rfc6761.
+        SPECIAL_USE_DOMAIN_NAMES = [
+          /\A(.*\.)?localhost\z/,
+          /\A(.*\.)?test\z/,
+
+          # While +.dev.+ is not described https://tools.ietf.org/html/rfc6761,
+          # it's common enough for developers to host their apps at +dev.+
+          # domains.
+          /\A(.*\.)?dev\z/,
+        ]
+
+        def spoofed?(host)
+          host.present? &&
+            SPECIAL_USE_DOMAIN_NAMES.any? { |domain| domain === host } ||
+            ip_address?(host)
```

Comment:
@matthewd are you fine with this? I think every ip address in the `X-Forwarded-Host` will pass trough and can be spoof us if we allow them in this header.

Reply:
I don't understand this. The above conditions all sound like trusted hostnames to me?